### PR TITLE
8331560: Refactor Hotspot container detection code so that subsystem delegates to controllers

### PR DESCRIPTION
**Pull request by jerboaa**
Please review this preparatory PR which is an enabler for some bugfix/alignment work (e.g. [JDK-8322420](https://bugs.openjdk.org/browse/JDK-8322420)). The idea is to delegate limit lookup to controllers (`memory`, `cpu`, etc.) from the overarching `CgroupSubsystem` class. This way we can - once the cg type has been detected - "adjust" a controller's path to the limit files when the controller is being created (on init) and then left alone. In this case, the two preparatory entry points are `CgroupUtil::processor_count()` taking a version-agnostic `CgroupCpuController` to do the actual look-up and `CgroupMemoryController::read_memory_limit_in_bytes()` for the same in terms of memory limits.

This enables setting the contoller's path to the interface files (wherever it might be in the hierarchy of it's original cgroup path), look up the limit and "freeze" the path once it found a - lower - limit.

It also ensures that both cgroup versions return `-1` or `-2` (`OSCONTAINER_ERROR`) - in both cases negative - for some notion of unlimited. I.e. it makes the upper bound by the host's physical memory apparent in the version agnostic classes.

Testing:
- [x] GHA
- [x] gtest:cgroupTest tests
- [x] container tests on Linux cgv1 and cgv2 on x86_64

Thoughts?

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331560](https://bugs.openjdk.org/browse/JDK-8331560): Refactor Hotspot container detection code so that subsystem delegates to controllers (**Enhancement** - P3)

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [dcd0554b](https://git.openjdk.org/jdk/pull/19085/files/dcd0554b66cbdb57508b7fae79aa0de814e99a26)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19085/head:pull/19085` \
`$ git checkout pull/19085`

Update a local copy of the PR: \
`$ git checkout pull/19085` \
`$ git pull https://git.openjdk.org/jdk.git pull/19085/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19085`

View PR using the GUI difftool: \
`$ git pr show -t 19085`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19085.diff">https://git.openjdk.org/jdk/pull/19085.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19085#issuecomment-2095457740)

### CODE REVIEW

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp
Comment (jankratochvil) [inline src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp:139]:
```suggestion
  }
  do_trace_log(memlimit, phys_mem);
  return (jlong)memlimit;
```

**Comment by jankratochvil** [comment]
There is the diamond inheritance problem. I am providing a patch, it would be too much clicking in the Github review.
[diamond.patch.txt](https://github.com/openjdk/jdk/files/15403710/diamond.patch.txt)

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp
Comment (jankratochvil) [inline src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp:214]: `memory_swap_usage_impl` could be a method. Likewise for `mem_soft_limit_val`, `mem_swp_limit_val`, `mem_swp_current_val` and `mem_limit_val`.

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (jankratochvil) [inline src/hotspot/os/linux/cgroupSubsystem_linux.cpp:70]: Could this be a dtor?

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (jankratochvil) [inline src/hotspot/os/linux/cgroupSubsystem_linux.cpp:67]:
```suggestion
    CgroupV2MemoryController* memory = new CgroupV2MemoryController(cg_infos.memory._mount_path, cg_infos.memory._cgroup_path);
```
First could there be:
class CgroupInfos {
  CgroupInfo[5];
  CgroupInfo &cpuset = CgroupInfo[0];
  CgroupInfo &cpu = CgroupInfo[1];
  etc.
};
and remove:
#define CG_INFO_LENGTH 5
#define CPUSET_IDX     0
#define CPU_IDX        1
...
Second the constructor should have `const CgroupInfo &` as a parameter.

**Comment by jerboaa** [comment]
@jankratochvil Thanks for the reviews and suggestions. I'll have another think about how to improve this!

**Comment by jdksjolen** [comment]
Hi @jerboaa , you can ping me when you want another review :).

**Comment by tstuefe** [comment]
> Hi @jerboaa , you can ping me when you want another review :).

God, eager Reviewers, what a rarity :-)

**Comment by jerboaa** [comment]
> Hi @jerboaa , you can ping me when you want another review :).

Sure. Thanks!

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp
Comment (jerboaa) [reply src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp:132]: Fixed.

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp
Comment (jerboaa) [reply src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp:214]: While true those should intentionally be local to the compilation unit. Adding them as methods clutters the code needlessly, IMO.

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (jerboaa) [reply src/hotspot/os/linux/cgroupSubsystem_linux.cpp:70]: It could probably, but I'd like to keep this patch minimal. We could refactor `cg_info` handling in another patch of course.

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (jerboaa) [reply src/hotspot/os/linux/cgroupSubsystem_linux.cpp:67]: Feel free to propose this cleanup as a separate patch. I'd like to keep this one on-topic to do controller delegation if that's ok.

**Comment by jerboaa** [comment]
@jdksjolen Ready for your review when you get a chance. Thanks!

**Comment by jdksjolen** [comment]
> @jdksjolen Ready for your review when you get a chance. Thanks!

I'm hoping to get to this next week, just FYI!

**Comment by jerboaa** [comment]
GHA failure of [serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage](https://github.com/jerboaa/jdk/actions/runs/9290657797#user-content-serviceability_jvmti_objectmonitorusage_objectmonitorusage) on Linux 32 bit seems unrelated to this patch.

**Comment by jdksjolen** [comment]
> > @jdksjolen Ready for your review when you get a chance. Thanks!
> 
> I'm hoping to get to this next week, just FYI!

I forgot that I have a short week this week (Sweden's national day = holiday), so I didn't have time to review the code this week :-(.

**Comment by jerboaa** [comment]
@jdksjolen Hi! Perhaps you have time this week for a review? Thanks!

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp
Comment (tstuefe) [inline src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp:52]: Drive by comment. I don't understand the aim behind the complexity, but multiple inheritance, even with pure abstract side classes, is usually not used. It is also against the hotspot style guide:

```
Restrict inheritance to the "is-a" case; use composition rather than
non-is-a related inheritance.

Avoid multiple inheritance.  Never use virtual inheritance.
```

Comment (jdksjolen) [reply src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp:52]: This kind of multiple-inheritance isn't a big deal, it's the same as having multiple `implements` classes in Java. @jerboaa , is it possible to avoid the multiple inheritance by stuffing a `CgroupV1Controller` into the `CgroupV1MemoryController` and so on? Or at least *only* inheriting from classes which are abstract and **only** exposes a set of `virtual = 0;` type methods.

Basically this:

```c++
class CgroupV1MemoryController {
  CgroupV1Controller _parent;
};
```

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp:69]: Can you add `override` to the end of the virtual functions here? `jlong memory_usage_in_bytes() override;`, like so.

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp
Comment (jdksjolen) [reply src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp:52]: Reading this code a bit more, seems like `CgroupController` just exposes a virtual function `subsystem_path`. But this function is only ever called from within `CgroupController`, and we never seem to want to handle a Cgroup controller in a generic way anyway. We can just make the subsystem_path be an argument to all of the functions which require it in `CgroupController` and make that class `AllStatic` and have no one inherit from it.

I also find the general usage of `malloc` concerning. looking at the class `CgroupV1Controller`, this doesn't have a corresponding destructor even though it seems like it owns the memory which it has pointers to. This can just hold the memory directly, no pointers necessary.

**Review by jdksjolen — COMMENTED** [review]
This code has (in general, not just this PR) a lot of unnecessary complexity. We can easily shave that complexity down, there's a lot of low hanging fruit here it seems. I'm seriously wondering if maybe we should get to these easier parts before we do another refactoring.

**Comment by jerboaa** [comment]
Thanks for the reviews!

> This code has (in general, not just this PR) a lot of unnecessary complexity. We can easily shave that complexity down, there's a lot of low hanging fruit here it seems. I'm seriously wondering if maybe we should get to these easier parts before we do another refactoring.

Please explain what exactly you mean by "unnecessary complexity". Happy to shave off some complexity if we *can*.

The point of this patch is that we want an API (version agnostic) so as to be able to determine the path to the interface files where the lowest limit is set (since cgroups is hierarchical). The path to the interface files might be different for the memory controller (as compared to the cpu controller, for example). So in order to be able to do that we need to have a way to adjust the `subsystem_path()`, determine the limit on that path and either continue or stop (if the limit is lower than the one previously seen). Then once we've found the correct paths for each controller use the code as we used to. [JDK-8333446](https://bugs.openjdk.org/browse/JDK-8333446) adds tests where this fails currently and shows the problem.

That's the point of this patch. It adds those API points for `cpu` (`CgroupUtil::processor_count()`) and `memory` (`CgroupMemoryController::read_memory_limit_in_bytes()`). Hope that makes it clearer.

Some of the other complexity is totally orthogonal to this patch. For example, caching a value for a short time rather than reading from the interface file. See [JDK-8232207](https://bugs.openjdk.org/browse/JDK-8232207) for details.

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp
Comment (jerboaa) [reply src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp:52]: Basically a `CgroupController`'s job is to read from the interface file. In order to do so it needs a base path to the interface files, which is determined elsewhere.

I'll rework the patch to get rid of the multiple-inheritance.

File: src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp:53]: Style: `char*`

File: src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp:62]: Style: `char*`

File: src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp:89]: Style: hug the asterisk to the name

Pre-existing: Please initialize the members in intiializer list.

File: src/hotspot/os/linux/cgroupSubsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupSubsystem_linux.hpp:185]: We always expect `T` to be a `T*`, so make it so.

Change:

```c++
T* _controller;
```

and remove the `*` from all instantiations.

File: src/hotspot/os/linux/cgroupSubsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupSubsystem_linux.hpp:195]: `T*`

File: src/hotspot/os/linux/cgroupSubsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupSubsystem_linux.hpp:189]: `T*`

**Review by jdksjolen — COMMENTED** [review]
OK, I'm getting a handle of this change. Some smaller review comments. I'll give it another round soon.

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp:93]: Should it be called trace when we log on a debug level?

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp:56]: It seems simpler to me to have this as a `CgroupV1Controller` and perhaps have a `reader()` accessor that returns a pointer. This class currently has a memory leak as the malloc-allocated `_reader` is never freed in the destructor. Up to you what you want to do.

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp:53]: Do we expect this class to be subclassed later on/in the future? If not, then we can declare it final as so:

```c++
class CgroupV1MemoryController final : public CgroupMemoryController
```

File: src/hotspot/os/linux/cgroupSubsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupSubsystem_linux.hpp:217]: `= 0`?

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp:144]: Pre-existing: No destructor freeing these.

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp
Comment (jerboaa) [reply src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp]: renamed to `verbose_log`

File: src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp
Comment (jerboaa) [reply src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp:157]: A `CgroupSubsystem` instance lives as long as the JVM lives. I'm not sure this is worth the extra complexity. For example, I'd rather avoid copying those path strings around and free them again (`CgroupV1Controller` and `CgroupV2Controller` destructors don't free, for example, because they get re-used latter avoiding this extra copying around). Instead essentially use those singletons and let the OS cleanup handle de-allocation.

**Comment by jerboaa** [comment]
Thanks for the review. Updated the patch!

File: src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp:77]: Take this as a reference to avoid needless copying.

File: src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp:107]: @jerboaa, Github doesn't let me comment on unchanged lines, so I'm not trying to talk about this line, but right beneath it :-). There are a lot of methods which are still declared which I think have been moved to the `CgroupV2MemoryController`, and so these should be removed, right?

File: src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp
Comment (jdksjolen) [inline src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp:42]: Style, pre-existing: `char *` -> `char*`.

**Review by jdksjolen — COMMENTED** [review]
Hi! Some more review comments.

**Comment by jerboaa** [comment]
Updated the patch with your suggestions. Thanks!

**Review by tstuefe — APPROVED** [review]
This has been shaping up nicely. From my point of view, this is good to go. Leave it up to @jdksjolen to add any more suggestions.

**Comment by jerboaa** [comment]
Thanks for the review, Thomas!

**Review by jdksjolen — APPROVED** [review]
I went through it again, LGTM. Nice!

**Comment by jerboaa** [comment]
Thanks for the reviews!

**Pull request closed without merging**
